pkg/frontend/querymiddleware/spin_off_subqueries.go

```diff
@@ -0,0 +1,315 @@
+
+	queryable := newSpinOffSubqueriesQueryable(req, annotationAccumulator, s.next, s.rangeHandler)
+
+	qry, err := newQuery(ctx, req, s.engine, lazyquery.NewLazyQueryable(queryable))
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to create new query from subquery spin request", "err", err)
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	res := qry.Exec(ctx)
+	extracted, err := promqlResultToSamples(res)
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to execute spun off subquery", "err", err)
+		return nil, mapEngineError(err)
+	}
+
+	// Note that the positions based on the original query may be wrong as the rewritten
+	// query which is actually used is different, but the user does not see the rewritten
+	// query, so we pass in an empty string as the query so the positions will be hidden.
+	warn, info := res.Warnings.AsStrings("", 0, 0)
+
+	// Add any annotations returned by the sharded queries, and remove any duplicates.
+	// We remove any position information for the same reason as above: the position information
+	// relates to the rewritten expression sent to queriers, not the original expression provided by the user.
+	accumulatedWarnings, accumulatedInfos := annotationAccumulator.getAll()
+	warn = append(warn, removeAllAnnotationPositionInformation(accumulatedWarnings)...)
+	info = append(info, removeAllAnnotationPositionInformation(accumulatedInfos)...)
+	warn = removeDuplicates(warn)
+	info = removeDuplicates(info)
+
+	return &PrometheusResponse{
+		Status: statusSuccess,
+		Data: &PrometheusData{
+			ResultType: string(res.Value.Type()),
+			Result:     extracted,
+		},
+		Headers:  queryable.getResponseHeaders(),
+		Warnings: warn,
+		Infos:    info,
+	}, nil
+}
+
+func newSpinOffQueryHandler(codec Codec, logger log.Logger, sendURL string) (MetricsQueryHandler, error) {
+	rangeQueryURL, err := url.Parse(sendURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid spin off URL: %w", err)
+	}
+
+	return &spinOffQueryHandler{
+		codec:         codec,
+		logger:        logger,
+		rangeQueryURL: rangeQueryURL,
+	}, nil
+}
+
+// spinOffQueryHandler is a query handler that takes a request and sends it to a remote endpoint.
+type spinOffQueryHandler struct {
+	codec         Codec
+	logger        log.Logger
+	rangeQueryURL *url.URL
+}
+
+func (s *spinOffQueryHandler) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	httpReq, err := s.codec.EncodeMetricsQueryRequest(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding request: %w", err)
+	}
+	httpReq.RequestURI = "" // Reset RequestURI to force URL to be used in the request.
+	// Override the URL with the configured range query URL (only parts that are set).
+	if s.rangeQueryURL.Scheme != "" {
+		httpReq.URL.Scheme = s.rangeQueryURL.Scheme
+	}
+	if s.rangeQueryURL.Host != "" {
+		httpReq.URL.Host = s.rangeQueryURL.Host
+	}
+	if s.rangeQueryURL.Path != "" {
+		httpReq.URL.Path = s.rangeQueryURL.Path
+	}
```

Comment:
This behaviour seems a bit confusing to me - why not just always use the values that have been configured? 

For example, if the configured URL is `http://foo`, then we'll always preserve the path from the original request, but this may not be what the user intended - they may want to target the root of `foo`.